### PR TITLE
New feature - autogeneration of cards or TSV-configs based on a list of words.

### PR DESCRIPTION
**Pull request by Otter-man**
So, I finished my feature for mnemocards. 
Sorry, I didn't touch TO-DO you mentioned in our previous communication, but next, I would probably be working on making a site for automatic creation of cards from words (and maybe a visual card editor), with mnemocards as the back-end, and maybe will work on things you mentioned along the way.

New features allow you to generate apkg or TSV-files from a list of words. It uses [py-googletrans library](https://github.com/ssut/py-googletrans) to generate translations for words and phrases.

To handle the creation of this new type of cards I wrote a new type for config - "autogenerate"
I also wrote two new commands:
* `maketsv` to create tsv-files based on translation, so you could manually change some values for columns before generating cards.
* `id` to generate random UUID. I used it to generate UUID for my custom decks and I think it could be valuable to other people who use `mnemocards`.

Details on how to use this are in the updated [README.md](README.md)
You can also check files in the example folder - I added "gtrans_generated" folder there with two configs that demonstrate basic use cases.

Also, I too studied Japanese in the past and so made a specific effort to add automatic generation of furigana for automatically generated cards with [pykakasi library](https://github.com/miurahr/pykakasi). It supports three types of furigana (written hiragana, katakana, and romaji). It's not as precise as hand-made furigana (it doesn't add furigana to each kanji, treating multi-kanji words as a whole instead) but it still can be very useful for studying Japanese. More in the [README.md](README.md)

I also wrote functional tests for new features and most new code is also covered by unit-tests. 

Your code was very well structured, it was a joy to built things on top of it.

HMU if you have any questions, hope you review this and will approve my contribution to the project.

### CODE REVIEW

**Comment by guiferviz** [comment]
Hi @Otter-man.
I haven't read you yet, it will take me a while to go through it but I will do it for sure. Thank you very much for your contribution!!!

File: README.md [hunk 1]
```diff
@@ -339,6 +376,62 @@ learning, you should specify here the language.
...
+This flag works the same as in `vocabulary` type, but generates furigana 
+automatically.
+This flag only works if the `lang:original` is set to japanese (`ja`). 
+If furigana is activated, pronunciation from Google Translate will be removed
+from the card.
+By default false.
+* `furigana_type`.
+You can choose what type of transliteration to use for furigana. Avaiable values
+`hira` for hiragana (this is set by default if you skip this flag), `kana` for
+katakana and `hepburn` for romaji.
```
Comment (guiferviz) [inline README.md:422]: :astonished: really cool.

**Comment by guiferviz** [comment]
Really cool @Otter-man!!! Thanks for this contribution.
I've tested it following your detailed instructions of the README file and it works perfect!
I just left some minor comments, I will be really happy to merge your code after you review them.

Can you also provide me the command that I need to run the tests? I'm getting some errors and I'm sure I'm missing something:
image
I've tried running the pytest from the root dir of the project and from the tests folder, but it fails from both of them.

Thanks again and sorry about taking so much time before coming back to you!

**Comment by Otter-man** [comment]
@guiferviz hi!

About tests: I was careless and made tests dependant on the project folder being "mnemocards" so they didn't work for you, since you named fork folder after me 😅 Please download my latest commits and try to run "pytest test_testing.py" from the /tests folder - now test only depend on you running pytest from /tests folder.

I also fixed some tests for translations. Since google change translations for words from time to time, tests now look not for the full match but only that at least 2 out of 4 example words matches.
Also, some tests showed that google translate API answers changed a little, so I also fixed it.

In other words, just check out tests, everything should pass correctly. 

And sorry, I don't see any comments you mentioned for me to check (except for the furigana generator), could you point them to me?

File: mnemocards/maketsv.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+                translations.append(one_translation)
+        if ip_block:
+            print('''Too many requests to GoogleAPI, your IP is blocked for 1 hour.
+Proceeding generation of cards from already translated words.
+Please run generation for remainding words again in an hour. ''')
+            continue
+
+        if words.index(word_block) != len(words) - 1:
+            print(
+                f"{len(translations)} words translated. Waiting 180 seconds for next translation")
```
[the remark this answers is not in the document]
Comment (Otter-man) [reply mnemocards/maketsv.py:55]: I spent several hours testing it with proxies and it something near 25 (or something near) requests per 150-180 second and then the limitation is reset. As far as I  tested, if you wait 3 minutes between 25-word batches you never get blocked and can use Google Translate API as much as you want.

**Comment by guiferviz** [comment]
@Otter-man sorry, I didn't submit my review, you should be able to see my comments now.

Thanks! Tests are working now. Great contribution, I know tests are important even if I didn't do any... :)

It's not a good idea to make the result of the tests depend on what the Google API returns, it would be better to mock the API call instead of using that "2 out of 4" technique. You don't need to change it anyway, it's fine for me.

File: requirements.in [hunk 3]
```diff
@@ -6,7 +6,9 @@ iso-639
 pygithub
 markdown2
 Pygments
-
+pykakasi
+googletrans==3.1.0a0
+chardet==3.0.4
```
[the remark this answers is not in the document]
Comment (Otter-man) [reply requirements.in:11]: Yes! Both versions are specific for a reason!
The latest version of googletrans library has a fifty-fifty chance of not working when translating in batches - [this issue](https://github.com/ssut/py-googletrans/issues/234). And generally, I found the 3.1.0a0 version to be more stable, and it's enough for our use.

About chardet - two different libraries have it in dependencies (sorry, don't remember which), but one specifies a later version of chardet when installed (but still works fine if you manually install 3.0.4 - it seems it needs chardet for some functions that aren't used by mnemocards). 
But if you have a later version of chardet installed mnemocards crashes for some reason, so I had to make a crutch and specify `chardet==3.0.4` in requirements to overwrite a dependency of later version.

File: mnemocards/maketsv.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Otter-man) [reply mnemocards/maketsv.py:55]: I got to test it when I migrated my 1000+ words deck. It just runs for a few hours but you never get 1-hour block.

**Comment by Otter-man** [comment]
I don't know how to mock API call in this case since one of the reasons for the tests is precisely to check if the API returns haven't changed. This test already helped me today to catch API change and fix it.

**Comment by guiferviz** [comment]
From your previous comment I've got that the output translations were different, not that the schema of the response was different.
I'm not saying that you need to do it, but from my point of view the best approach for testing this is:

* Mock the calls and check that you can build the apkg.
* Create a new test that checks that the API output is what you are expecting. For example, if you are expecting a JSON with two properties "lang" and "translation", check that you have those fields in the response of the API but do not check the exact translation as we can assume that Google Translator will return the best one.

File: requirements.in (hunk 3 above)
[the remark this answers is not in the document]
Comment (guiferviz) [reply requirements.in:11]: Ok, thanks! Let's leave the fixed versions then.

File: mnemocards/maketsv.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (guiferviz) [reply mnemocards/maketsv.py:55]: Amazing, thanks.

**Comment by guiferviz** [comment]
Happy to merge, should I close it or do you still want to add something else?

**Comment by Otter-man** [comment]
@guiferviz yes, merge away, I won't be adding anything else right now.

About the tests: API answers don't have a defined structure, most of the data are in unnamed lists of lists of varying length. Due to that, it's hard to detect when structure changed and test only for that. 
I see how I can probably try to define how the structure shouldn't be, but for now, I'm not doing it. 

I'm still working on mnemocards-as-web-service version and so I probably will update tests in the future and will make a separate pull-request for stuff that can be useful in this repo!

**Comment by guiferviz** [comment]
Merged! Thank you so much for this contribution @Otter-man
I will be really happy to merge more changes from you. As I said initially, mnemocards already meets my requirements and I'm a bit busy lately, but if you need a new feature maybe we could collaborate together on something. You know where to find me :) Thanks again!
